### PR TITLE
Enforce hex and city restrictions for token abilities

### LINKED ISSUE #1228: [1846] C&WI private allows incorrect station laying
In 1846 the C&WI private allows laying of an incorrect station in all Chicago, not just in the reserved spot.

### LINKED ISSUE #1461: 1846: illegal upgrade tile
https://18xx.games/game/6760?action=77 

in this action GT place a tile in D8 that is not legal.
According to rules "6.47 Connection: Some of the newly placed track of a non-city tile
must connect to a token of the corporation placing it. "

D8

### PR DESCRIPTION
**Pull request by cheesinglee**
Also includes a little fix-up for the token log messages for cases where the hex has no location name.

Fixes #1228
Fixes #1461

### CODE REVIEW

File: lib/engine/step/special_token.rb [hunk 1]
```diff
@@ -22,13 +23,20 @@ def blocks?
...
+
+        if token_ability.check_target(action.city)
+          place_token(
+            entity.owner,
+            action.city,
+            action.token,
+            teleport: token_ability.teleport_price,
+          )
+        else
+          raise GameError,
```
Comment (tobymao) [inline lib/engine/step/special_token.rb:36]: this should be @game.gate_error(...)

File: lib/engine/step/tokener.rb [hunk 2]
```diff
@@ -29,22 +29,23 @@ def place_token(entity, city, token, teleport: false)
...
+        location_log = "(#{hex.location_name})" if hex.location_name
 
         case token.type
         when :neutral
           entity.tokens.delete(token)
           token.corporation.tokens << token
           @log << "#{entity.name} places a neutral token on #{hex.name}#{price_log}"
         else
-          @log << "#{entity.name} places a token on #{hex.name} (#{hex.location_name})#{price_log}"
+          @log << "#{entity.name} places a token on #{hex.name} #{location_log}#{price_log}"
```
Comment (tobymao) [inline lib/engine/step/tokener.rb:48]: this is fine, but i think we should also add in config/game/g_1846 chicago as the location name

File: lib/engine/ability/token.rb [hunk 3]
```diff
@@ -5,13 +5,20 @@
...
-      def setup(hexes:, price:, teleport_price: nil, extra: nil)
+      def setup(hexes:, price:, teleport_price: nil, extra: nil, city: nil)
         @hexes = hexes
         @price = price
         @teleport_price = teleport_price
         @extra = extra || false
+        @city = city
+      end
+
+      def check_target(city)
```
Comment (tobymao) [inline lib/engine/ability/token.rb:18]: maybe it should be check_city

File: lib/engine/ability/token.rb [hunk 4]
```diff
@@ -5,13 +5,20 @@
...
+        match_hex_id = @hexes.include?(city.hex.id)
+        match_city = @city.nil? || city == city.tile.cities[@city]
+        match_hex_id && match_city
```
Comment (tobymao) [inline lib/engine/ability/token.rb:21]: @hexes.include?(city.hex.id) && (!@city || city == city.tile.cityl[@city])

**Comment by cheesinglee** [comment]
Addressed comments, and also pushed a change the fixes #1461

File: lib/engine/step/special_token.rb [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 require_relative 'base'
 require_relative 'tokener'
+require_relative '../game_error'
```
Comment (tobymao) [inline lib/engine/step/special_token.rb:5]: this doesn't need to be required

Comment (tobymao) [reply lib/engine/step/special_token.rb:5]: you also spelled game_error wrong as gate_error

Comment (tobymao) [reply lib/engine/step/special_token.rb:5]: and a typo of can not

**Comment by tobymao** [comment]
i've made some cleanups in the next commit

**Pull request merged**